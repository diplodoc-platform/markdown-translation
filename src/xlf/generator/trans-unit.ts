import {XMLBuilder} from 'fast-xml-parser';

const options = {format: true, ignoreAttributes: false};
const builder = new XMLBuilder(options);

export type TransUnitParameters = {
    indentation?: number;
    target?: string;
    source?: string;
    id: number;
};

export type TransUnitData = {
    'trans-unit': {
        source?: XLFHasText;
        target?: XLFHasText;
    } & XLFHasID;
};

export type XLFHasText = {
    '#text'?: string;
};

export type XLFHasID = {
    '@_id'?: number;
};

function generate({source, target, id, indentation = 0}: TransUnitParameters) {
    const data: TransUnitData = {
        'trans-unit': {
            '@_id': id,
        },
    };

    if (source?.length) {
        data['trans-unit'].source = {
            '#text': source,
        };
    }

    if (target?.length) {
        data['trans-unit'].target = {
            '#text': target,
        };
    }

    return builder
        .build(data)
        .split('\n')
        .filter(Boolean)
        .map((s: string) => ' '.repeat(indentation + 2) + s)
        .join('\n');
}

export {generate};
export default {generate};
