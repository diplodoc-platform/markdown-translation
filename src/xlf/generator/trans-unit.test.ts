import {generate} from './trans-unit';

describe('trans-unit generator', () => {
    it('generates trans-unit with provided: source, id', () => {
        const generated = generate({source: 'Sentence about something', id: 1});

        expect(generated).toMatchSnapshot();
    });

    it('generates trans-unit with provided source, id, target', () => {
        const generated = generate({
            source: 'Sentence about something',
            target: 'Предложение о чем-то',
            id: 1,
        });

        expect(generated).toMatchSnapshot();
    });
});
