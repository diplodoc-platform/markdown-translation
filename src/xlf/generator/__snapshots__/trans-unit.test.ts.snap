// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`trans-unit generator generates trans-unit with provided source, id, target 1`] = `
"  <trans-unit id=\\"1\\">
    <source>Sentence about something</source>
    <target>Предложение о чем-то</target>
  </trans-unit>"
`;

exports[`trans-unit generator generates trans-unit with provided: source, id 1`] = `
"  <trans-unit id=\\"1\\">
    <source>Sentence about something</source>
  </trans-unit>"
`;
